wallet/createtx.go

```diff
@@ -109,148 +109,159 @@ func (s secretSource) GetScript(addr btcutil.Address) ([]byte, error) {
 func (w *Wallet) txToOutputs(outputs []*wire.TxOut, keyScope *waddrmgr.KeyScope,
 	account uint32, minconf int32, feeSatPerKb btcutil.Amount,
 	coinSelectionStrategy CoinSelectionStrategy, dryRun bool) (
-	tx *txauthor.AuthoredTx, err error) {
+	*txauthor.AuthoredTx, error) {
 
 	chainClient, err := w.requireChainClient()
 	if err != nil {
 		return nil, err
 	}
 
-	dbtx, err := w.db.BeginReadWriteTx()
-	if err != nil {
-		return nil, err
-	}
-	defer func() { _ = dbtx.Rollback() }()
-
-	addrmgrNs, changeSource, err := w.addrMgrWithChangeSource(
-		dbtx, keyScope, account,
-	)
-	if err != nil {
-		return nil, err
-	}
-
 	// Get current block's height and hash.
 	bs, err := chainClient.BlockStamp()
 	if err != nil {
 		return nil, err
 	}
 
-	eligible, err := w.findEligibleOutputs(
-		dbtx, keyScope, account, minconf, bs,
-	)
-	if err != nil {
-		return nil, err
-	}
-
-	var inputSource txauthor.InputSource
-
-	switch coinSelectionStrategy {
-	// Pick largest outputs first.
-	case CoinSelectionLargest:
-		sort.Sort(sort.Reverse(byAmount(eligible)))
-		inputSource = makeInputSource(eligible)
-
-	// Select coins at random. This prevents the creation of ever smaller
-	// utxos over time that may never become economical to spend.
-	case CoinSelectionRandom:
-		// Skip inputs that do not raise the total transaction output
-		// value at the requested fee rate.
-		var positivelyYielding []wtxmgr.Credit
-		for _, output := range eligible {
-			output := output
-
-			if !inputYieldsPositively(&output, feeSatPerKb) {
-				continue
-			}
-
-			positivelyYielding = append(positivelyYielding, output)
+	var tx *txauthor.AuthoredTx
+	err = walletdb.Update(w.db, func(dbtx walletdb.ReadWriteTx) error {
+		addrmgrNs, changeSource, err := w.addrMgrWithChangeSource(
```

Comment:
This doesn't look thread-safe, but from our offline discussion I understand that etcd no longer guarantees that this code isn't executed concurrently like it was the case for bbolt.

I think this applies to other changes too.

Reply:
Yes, one key thing about using the `View` and `Update` closures is that we only guarantee that transactions are serializable, but there's no guarantee of a single `Update` at a time. Also the closure may retry so any logic inside has to be aware of this.

Reply:
Echoing the offline discussion here: my suggestion would be to for now use the familiar global write lock with etcd that bbolt uses too to avoid the risk of breaking anything. The etcd backend plugin itself is probably more than enough risk already, let alone if the app code would need to be changed too. Data safety above performance.

Reply:
Discussed this a bit more offline. The current plan is to add a lock specifically for the `etcd` DB instance that uses the `walletdb` to avoid concurrency issues. If this slows us down too much in `lnd` then we might need to re-architecture the wallet somewhat to support the reset callbacks and account for proper thread safety.

Reply:
Are you sure that the other database instances are completely safe to use without the lock? As a user, my preference would be to wrap them all in a lock for the first release. Lightning data is critical.

Reply:
> Are you sure that the other database instances are completely safe to use without the lock?

No, not yet. This will need to be made sure and is being discussed in other PRs.
But assuming that there is a global lock for the wallet DB, can we proceed with the PR as is?

Reply:
Yes looks good then.

Reply:
> This doesn't look thread-safe and from our offline discussion I understand that etcd no longer guarantees that this code isn't executed concurrently like it was the case for bbolt.

Check out `txCreator` this only ever runs serially.